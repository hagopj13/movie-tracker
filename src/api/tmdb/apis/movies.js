// @flow
import axios from '../axiosInstance';

export type GetDiscoverMoviesParams = {
  page?: number,
  sort_by: string,
  'vote_count.gte'?: number,
  with_genres?: string,
  'primary_release_date.gte'?: string,
  'primary_release_date.lte'?: string,
};
export const getDiscoverMovies = (params: GetDiscoverMoviesParams) =>
  axios.get('/discover/movie', { params });

export type GetUpcomingMoviesParams = {
  page?: number,
};
export const getUpcomingMovies = (params: GetUpcomingMoviesParams) =>
  axios.get('/movie/upcoming', { params });

export type SearchMoviesParams = {
  page?: number,
  query: string,
};
export const searchMovies = (params: SearchMoviesParams) => axios.get('/search/movie', { params });

export const getMovieDetails = (id: string, sessionId?: string) =>
  axios.get(`/movie/${id}`, {
    params: {
      session_id: sessionId,
      append_to_response: 'credits,keywords,reviews,recommendations,account_states',
    },
  });

export const getMovieUserState = (id: string, sessionId: string) =>
  axios.get(`/movie/${id}/account_states?session_id=${sessionId}`);

export const setIsMovieFavorite = (
  id: string,
  isFavorite: boolean,
  sessionId: string,
  accountId: string,
) =>
  axios.post(`/account/${accountId}/favorite?session_id=${sessionId}`, {
    media_type: 'movie',
    media_id: id,
    favorite: isFavorite,
  });

export const setIsMovieInWatchlist = (
  id: string,
  isInWatchlist: boolean,
  sessionId: string,
  accountId: string,
) =>
  axios.post(`/account/${accountId}/watchlist?session_id=${sessionId}`, {
    media_type: 'movie',
    media_id: id,
    watchlist: isInWatchlist,
  });

export const rateMovie = (id: string, rating: number, sessionId: string) =>
  axios.post(`/movie/${id}/rating?session_id=${sessionId}`, { value: rating });

export const deleteMovieRating = (id: string, sessionId: string) =>
  axios.delete(`/movie/${id}/rating?session_id=${sessionId}`);
