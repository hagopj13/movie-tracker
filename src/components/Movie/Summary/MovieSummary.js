// @flow
import React from 'react';
import moment from 'moment';
import { makeStyles, styled } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import Rating from '@material-ui/lab/Rating';

import type { Movie } from 'types';

type Props = {
  movie: Movie,
};

const SummarySection = styled(Box)(({ theme }) => ({
  marginTop: theme.spacing(1),
  [theme.breakpoints.down('xs')]: {
    textAlign: 'center',
  },
}));

const useStyles = makeStyles((theme) => ({
  root: {
    [theme.breakpoints.down('xs')]: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
    },
  },
  tagline: {
    marginTop: 0,
  },
  iconEmpty: {
    color: theme.palette.grey[500],
  },
}));

const MovieSummary = (props: Props) => {
  const { movie } = props;

  const classes = useStyles();

  const getReleaseYear = () => moment(movie.releaseDate).get('year');

  const renderTitle = () => (
    <SummarySection>
      <Typography>
        <Box component="span" fontWeight="fontWeightMedium" fontSize={34}>
          {movie.title}
        </Box>
        {movie.releaseDate && (
          <Box component="span" fontWeight="fontWeightLight" fontSize={34}>
            {` (${getReleaseYear()})`}
          </Box>
        )}
      </Typography>
    </SummarySection>
  );

  const renderTagline = () =>
    movie.tagline && (
      <SummarySection className={classes.tagline}>
        <Typography className={classes.tagline} variant="h6">
          <Box fontStyle="italic" fontWeight="fontWeightLight">
            {movie.tagline}
          </Box>
        </Typography>
      </SummarySection>
    );

  const renderRating = () =>
    movie.voteCount ? (
      <SummarySection>
        <Box display="flex" alignItems="center">
          <Rating
            classes={{ iconEmpty: classes.iconEmpty }}
            value={movie.voteAverage}
            precision={0.5}
            readOnly
          />
          <Box component="span" ml={0.75} fontWeight="fontWeightMedium" fontSize="h6.fontSize">
            {movie.voteAverage}
          </Box>
        </Box>
      </SummarySection>
    ) : null;

  const renderGenres = () =>
    movie.genres?.length > 0 ? (
      <SummarySection>
        <Typography variant="h6">
          <Box component="span">Genres: </Box>
          <Box component="span" fontWeight="fontWeightLight">
            {movie.genres.map((genre) => genre.name).join(', ')}
          </Box>
        </Typography>
      </SummarySection>
    ) : null;

  const renderRuntime = () =>
    movie.runtime ? (
      <SummarySection>
        <Typography variant="h6">
          <Box component="span">Runtime: </Box>
          <Box component="span" fontWeight="fontWeightLight">
            {`${Math.floor(movie.runtime / 60)}h ${movie.runtime % 60}m`}
          </Box>
        </Typography>
      </SummarySection>
    ) : null;

  const renderReleaseDate = () => (
    <SummarySection>
      <Typography variant="h6">
        <Box component="span">Release Date: </Box>
        <Box component="span" fontWeight="fontWeightLight">
          {movie.releaseDate ? moment(movie.releaseDate).format('MMM D, YYYY') : 'Unknown'}
        </Box>
      </Typography>
    </SummarySection>
  );

  const renderReleaseStatus = () => (
    <SummarySection>
      <Typography variant="h6">
        <Box component="span">Status: </Box>
        <Box component="span" fontWeight="fontWeightLight">
          {movie.status || 'Unknown'}
        </Box>
      </Typography>
    </SummarySection>
  );

  const renderOverview = () =>
    movie.overview && (
      <SummarySection>
        <Typography variant="h6">Overview</Typography>
        <Typography variant="h6">
          <Box fontWeight="fontWeightLight">{movie.overview}</Box>
        </Typography>
      </SummarySection>
    );

  return (
    <div className={classes.root}>
      {renderTitle()}
      {renderTagline()}
      {renderRating()}
      {renderGenres()}
      {renderRuntime()}
      {renderReleaseDate()}
      {renderReleaseStatus()}
      {renderOverview()}
    </div>
  );
};

export default MovieSummary;
