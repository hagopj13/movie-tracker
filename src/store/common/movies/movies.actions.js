// @flow
import typeGenerator from 'store/common/utils/typeGenerator';
import actionCreatorGenerator from 'store/common/utils/actionCreatorGenerator';
import type { Movie, Results } from 'types';

import MoviesActionTypes from './movies.types';

const fetchMovies = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES),
});

const fetchMoviesStart = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_START),
});

const fetchMoviesSuccess = (namespace) => (results: Results<Movie>) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_SUCCESS),
  payload: { results },
});

const fetchMoviesFailure = (namespace) => (error: string) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_FAILURE),
  payload: { error },
});

const fetchMoreMovies = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES),
});

const fetchMoreMoviesStart = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_START),
});

const fetchMoreMoviesSuccess = (namespace) => (results: Results<Movie>) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_SUCCESS),
  payload: { results },
});

const fetchMoreMoviesFailure = (namespace) => (error: string) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_FAILURE),
  payload: { error },
});

type Action = {
  type: string,
  payload?: any,
};

type MoviesActionCreators = {
  fetchMovies: () => Action,
  fetchMoviesStart: () => Action,
  fetchMoviesSuccess: (results: Results<Movie>) => Action,
  fetchMoviesFailure: (error: string) => Action,
  fetchMoreMovies: () => Action,
  fetchMoreMoviesStart: () => Action,
  fetchMoreMoviesSuccess: (results: Results<Movie>) => Action,
  fetchMoreMoviesFailure: (error: string) => Action,
};

export default (namespace: string): MoviesActionCreators =>
  actionCreatorGenerator<MoviesActionCreators>(namespace, {
    fetchMovies,
    fetchMoviesStart,
    fetchMoviesSuccess,
    fetchMoviesFailure,
    fetchMoreMovies,
    fetchMoreMoviesStart,
    fetchMoreMoviesSuccess,
    fetchMoreMoviesFailure,
  });
