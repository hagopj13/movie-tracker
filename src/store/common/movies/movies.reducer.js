// @flow
import reducerGenerator from 'store/common/utils/reducerGenerator';
import type { Movie } from 'types';

import MoviesActionTypes from './movies.types';

type State = {
  list: Movie[],
  pagination: {
    page: number,
    totalPages: number,
    totalResults: number,
  } | null,
};

export const defaultInitialState: State = {
  list: [],
  pagination: null,
};

const fetchMoviesStart = () => defaultInitialState;

const fetchMoviesSuccess = (state: State, action) => ({
  ...state,
  list: action.payload.results.list,
  pagination: action.payload.results.pagination,
});

const fetchMoreMoviesSuccess = (state: State, action) => ({
  ...state,
  list: state.list.concat(action.payload.results.list),
  pagination: {
    ...state.pagination,
    page: action.payload.results.pagination.page,
  },
});

const moviesActionHandler = {
  [MoviesActionTypes.FETCH_MOVIES_START]: fetchMoviesStart,
  [MoviesActionTypes.FETCH_MOVIES_SUCCESS]: fetchMoviesSuccess,
  [MoviesActionTypes.FETCH_MORE_MOVIES_SUCCESS]: fetchMoreMoviesSuccess,
};

export default (namespace: string, initialState?: State) =>
  reducerGenerator(namespace, moviesActionHandler, initialState ?? defaultInitialState);
