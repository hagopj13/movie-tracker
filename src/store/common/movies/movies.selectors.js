import { createSelector } from 'reselect';

const selectList = (state) => state.list;

const selectPagination = (state) => state.pagination;

const selectCurrentPage = createSelector([selectPagination], (pagination) => pagination?.page ?? 1);

const selectTotalPages = createSelector(
  [selectPagination],
  (pagination) => pagination?.totalPages ?? 0,
);

const selectTotalResults = createSelector(
  [selectPagination],
  (pagination) => pagination?.totalResults ?? 0,
);

export default {
  selectList,
  selectCurrentPage,
  selectTotalPages,
  selectTotalResults,
};
