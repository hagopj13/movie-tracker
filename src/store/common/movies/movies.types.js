export default {
  FETCH_MOVIES: 'FETCH_MOVIES',
  FETCH_MOVIES_START: 'FETCH_MOVIES_START',
  FETCH_MOVIES_SUCCESS: 'FETCH_MOVIES_SUCCESS',
  FETCH_MOVIES_FAILURE: 'FETCH_MOVIES_FAILURE',
  FETCH_MORE_MOVIES: 'FETCH_MORE_MOVIES',
  FETCH_MORE_MOVIES_START: 'FETCH_MORE_MOVIES_START',
  FETCH_MORE_MOVIES_SUCCESS: 'FETCH_MORE_MOVIES_SUCCESS',
  FETCH_MORE_MOVIES_FAILURE: 'FETCH_MORE_MOVIES_FAILURE',
};
