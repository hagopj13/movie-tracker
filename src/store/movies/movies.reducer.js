import { combineReducers } from 'redux';

import discoverReducer from './discover/discover.reducer';

export default combineReducers({
  discover: discoverReducer,
});
