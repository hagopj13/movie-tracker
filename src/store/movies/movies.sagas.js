import { all, call } from 'redux-saga/effects';

import discoverMoviesSagas from './discover/discover.sagas';

export default function* rootSaga() {
  yield all([call(discoverMoviesSagas)]);
}
