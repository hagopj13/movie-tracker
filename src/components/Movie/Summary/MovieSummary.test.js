import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import movies from 'store/fixtures/movies';

import MovieSummary from './MovieSummary';

describe('MovieSummary component', () => {
  let shallow;
  let movie;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    [movie] = movies.list;
    const mockProps = {
      movie,
    };
    wrapper = shallow(<MovieSummary {...mockProps} />);
  });

  it('should render the MovieSummary component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
