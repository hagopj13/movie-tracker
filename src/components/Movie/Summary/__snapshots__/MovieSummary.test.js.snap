// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieSummary component should render the MovieSummary component correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <Styled(Styled(MuiBox))>
    <WithStyles(ForwardRef(Typography))>
      <Styled(MuiBox)
        component="span"
        fontSize={34}
        fontWeight="fontWeightMedium"
      >
        Ad Astra
      </Styled(MuiBox)>
      <Styled(MuiBox)
        component="span"
        fontSize={34}
        fontWeight="fontWeightLight"
      >
         (2019)
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Typography))>
  </Styled(Styled(MuiBox))>
  <Styled(Styled(MuiBox))>
    <Styled(MuiBox)
      alignItems="center"
      display="flex"
    >
      <WithStyles(ForwardRef(Rating))
        classes={
          Object {
            "iconEmpty": "makeStyles-iconEmpty-3",
          }
        }
        precision={0.5}
        readOnly={true}
        value={3}
      />
      <Styled(MuiBox)
        component="span"
        fontSize="h6.fontSize"
        fontWeight="fontWeightMedium"
        ml={0.75}
      >
        3
      </Styled(MuiBox)>
    </Styled(MuiBox)>
  </Styled(Styled(MuiBox))>
  <Styled(Styled(MuiBox))>
    <WithStyles(ForwardRef(Typography))
      variant="h6"
    >
      <Styled(MuiBox)
        component="span"
      >
        Release Date: 
      </Styled(MuiBox)>
      <Styled(MuiBox)
        component="span"
        fontWeight="fontWeightLight"
      >
        Sep 17, 2019
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Typography))>
  </Styled(Styled(MuiBox))>
  <Styled(Styled(MuiBox))>
    <WithStyles(ForwardRef(Typography))
      variant="h6"
    >
      <Styled(MuiBox)
        component="span"
      >
        Status: 
      </Styled(MuiBox)>
      <Styled(MuiBox)
        component="span"
        fontWeight="fontWeightLight"
      >
        Unknown
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Typography))>
  </Styled(Styled(MuiBox))>
  <Styled(Styled(MuiBox))>
    <WithStyles(ForwardRef(Typography))
      variant="h6"
    >
      Overview
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="h6"
    >
      <Styled(MuiBox)
        fontWeight="fontWeightLight"
      >
        The near future, a time when both hope and hardships drive humanity to look to the stars and beyond. While a mysterious phenomenon menaces to destroy life on planet Earth, astronaut Roy McBride undertakes a mission across the immensity of space and its many perils to uncover the truth about a lost expedition that decades before boldly faced emptiness and silence in search of the unknown.
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Typography))>
  </Styled(Styled(MuiBox))>
</div>
`;
